File: backend/src/Modules/Masters/HMS.Modules.Masters/Domain/DiagnosticProvider.cs
using HMS.Shared.Kernel;

namespace HMS.Modules.Masters.Domain;

/// <summary>
/// An external organization a <see cref="DiagnosticService"/> can be outsourced to (e.g.
/// "Q-LAB") — part of the normalized replacement for the old flat DiagnosticTest.ReferenceLab
/// free-text string. Named generically ("Provider", not "Reference Lab") because Radiology can
/// outsource to an external imaging center just as Laboratory outsources to an external
/// pathology lab — this one entity covers both.
/// </summary>
internal class DiagnosticProvider : Entity
{
    public string Code { get; private set; } = null!;
    public string Name { get; private set; } = null!;
    public string? ContactDetails { get; private set; }
    public bool IsActive { get; private set; } = true;

    // Required by EF Core materialization.
    private DiagnosticProvider()
    {
    }

    private DiagnosticProvider(
        Guid id,
        string code,
        string name,
        string? contactDetails,
        bool isActive,
        Guid? createdBy)
        : base(id, createdBy)
    {
        Code = code;
        Name = name;
        ContactDetails = contactDetails;
        IsActive = isActive;
    }

    public static DiagnosticProvider Create(
        string code,
        string name,
        string? contactDetails,
        bool isActive,
        Guid? createdBy)
    {
        Guard.AgainstNullOrWhiteSpace(code, nameof(code));
        Guard.AgainstNullOrWhiteSpace(name, nameof(name));

        return new DiagnosticProvider(
            Guid.CreateVersion7(),
            code.Trim(),
            name.Trim(),
            contactDetails?.Trim(),
            isActive,
            createdBy);
    }

    public void Update(
        string code,
        string name,
        string? contactDetails,
        bool isActive,
        Guid? updatedBy)
    {
        Guard.AgainstNullOrWhiteSpace(code, nameof(code));
        Guard.AgainstNullOrWhiteSpace(name, nameof(name));

        Code = code.Trim();
        Name = name.Trim();
        ContactDetails = contactDetails?.Trim();
        IsActive = isActive;
        MarkUpdated(updatedBy);
    }
}

File: frontend/shared/dtos/diagnostics/diagnosticProvider.ts
/** Mirrors HMS.Modules.Masters.Contracts.DiagnosticProviderResponse — the "External Lab" from
 * Central Laboratory's UI, named generically since it also covers external imaging centers
 * for Radiology. See diagnosticCategory.ts for why this lives outside dtos/masters. */
export interface DiagnosticProvider {
  id: string;
  code: string;
  name: string;
  contactDetails?: string | null;
  isActive: boolean;
  createdAt: string;
  updatedAt?: string | null;
}

/** Mirrors HMS.Modules.Masters.Contracts.CreateDiagnosticProviderRequest. */
export interface CreateDiagnosticProviderRequest {
  code: string;
  name: string;
  contactDetails?: string | null;
  isActive: boolean;
}

/** Mirrors HMS.Modules.Masters.Contracts.UpdateDiagnosticProviderRequest. */
export interface UpdateDiagnosticProviderRequest {
  code: string;
  name: string;
  contactDetails?: string | null;
  isActive: boolean;
}

/** Mirrors HMS.Modules.Masters.Contracts.DiagnosticProviderListQuery. */
export interface DiagnosticProviderListQuery {
  page?: number;
  pageSize?: number;
  sort?: string;
  search?: string;
  isActive?: boolean;
}
